### PR TITLE
Jeffnye/experimental

### PR DESCRIPTION
**Pull request by jeffnye-gh**
Adds ZBB and more general extension mechanism. 

Extension enforcement is not complete but sufficient to enable/disable zbbb.

### CODE REVIEW

File: src/dromajo_main.cpp [hunk 1]
```diff
@@ -564,6 +566,15 @@ static BOOL net_poll_cb(void *arg) { return net_completed; }
 
 #endif
 
+static void usage_isa()
+{
+//    fprintf(dromajo_stderr,"\nSupported/Implemented ISA Extensions\n");
```
Comment (StanSofomoDev) [inline src/dromajo_main.cpp:571]: Whole body of this function is commented, is that intended?

File: include/dromajo_template.h [hunk 2]
```diff
@@ -55,6 +56,47 @@
...
+#define F3_SEXT     0x1
+
+#define HALF_WORD_MASK 0xFFFF
+
+#define CAPTURE_LOG    0
+#define REPORT_ILLEGAL 1
+
+#if REPORT_ILLEGAL == 1
+#define ILLEGAL_INSTR(S) { \
+  fprintf(dromajo_stderr,"HERE %s PC:0x%lx ENC:0x%08x\n",S,s->pc,insn); \
```
Comment (StanSofomoDev) [inline include/dromajo_template.h:81]: Is just `HERE` enough as a log message? Wouldn't something like `ILLEGAL INSTR HERE` be better?

File: include/dromajo_template.h [hunk 3]
```diff
@@ -82,15 +126,94 @@ static inline intx_t glue(rem, XLEN)(intx_t a, intx_t b) {
...
 static inline uintx_t glue(remu, XLEN)(uintx_t a, uintx_t b) {
     if (b == 0) {
         return a;
     } else {
         return a % b;
     }
 }
+// =========================================================================
+// =========================================================================
+//#if EN_ZBB == 1
```
Comment (StanSofomoDev) [inline include/dromajo_template.h:140]: Is this commented on purpose? I would add a TODO if it should be uncommented in the future.

File: tests/isa/run_one.sh [hunk 4]
```diff
@@ -0,0 +1,15 @@
+
+../../build/dromajo --ctrlc --trace 0 \
```
Comment (StanSofomoDev) [inline tests/isa/run_one.sh:2]: Might be good to allow it to take an argument for the benchmark file with `bmi_median.bare.riscv` being the default one.

File: tests/isa/tables.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
+def bin_to_hex(binary_str):
+    # Convert binary string to integer
+    decimal_value = int(binary_str, 2)
+    # Convert integer to hexadecimal and prefix with '0x'
+    hex_value = hex(decimal_value)
```
Comment (StanSofomoDev) [inline tests/isa/tables.py:5]: Does it automatically add consistend padding of hex values? If not, we could use `zfill` for that.

**Review by StanSofomoDev — APPROVED** [review]

File: tests/isa/run_one.sh (hunk 4 above)
Comment (jeffnye-gh) [reply tests/isa/run_one.sh]: old file, removed

File: tests/isa/tables.py (hunk 5 above)
Comment (jeffnye-gh) [reply tests/isa/tables.py:5]: noted, I'm going to leave it this way for now, so far it works as I intend.
I believe this utility eventually goes away. It is useful during development but is not a long term replacement for official documentation.

File: include/dromajo_template.h (hunk 3 above)
Comment (jeffnye-gh) [reply include/dromajo_template.h]: removed the dead code

File: include/dromajo_template.h (hunk 2 above)
Comment (jeffnye-gh) [reply include/dromajo_template.h]: corrected

File: src/dromajo_main.cpp (hunk 1 above)
Comment (jeffnye-gh) [reply src/dromajo_main.cpp]: fixed in push

File: src/dromajo_isa.cpp [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+    // Check if the segment starts with a valid base
+    std::string base;
+    if (segment.compare(0, 4, "rv32") == 0) {
+        flags.rv32 = true;
+        base = "rv32";
+    } else if (segment.compare(0, 4, "rv64") == 0) {
+        flags.rv32 = true;
+        base = "rv64";
+    } else if (segment.compare(0, 5, "rv128") == 0) {
+        flags.rv32 = true;
```
[the remark this answers is not in the document]
Comment (jeffnye-gh) [reply src/dromajo_isa.cpp]: yes another cut and paste issue

File: src/dromajo_isa.cpp [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+    flags.rv64  = false;
+    flags.rv128 = false;
+
+    // Check if the segment starts with a valid base
+    std::string base;
+    if (segment.compare(0, 4, "rv32") == 0) {
+        flags.rv32 = true;
+        base = "rv32";
+    } else if (segment.compare(0, 4, "rv64") == 0) {
+        flags.rv32 = true;
```
[the remark this answers is not in the document]
Comment (jeffnye-gh) [reply src/dromajo_isa.cpp]: yes this is a mistake, corrected.

**Pull request merged**